# UFC Predictions

### Step 1: Load Data

The following CSV files were retrieved from the Kaggle UFC dataset.

In [11]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

pd.options.display.max_columns = 999

dat_df = pd.read_csv("https://raw.githubusercontent.com/ekoly/DS-Unit-2-Applied-Modeling/master/data/ufc/data.csv")
prep_dat_df = pd.read_csv("https://raw.githubusercontent.com/ekoly/DS-Unit-2-Applied-Modeling/master/data/ufc/preprocessed_data.csv")
raw_fighters_df = pd.read_csv("https://raw.githubusercontent.com/ekoly/DS-Unit-2-Applied-Modeling/master/data/ufc/raw_fighter_details.csv")
raw_dat_df = pd.read_csv("https://raw.githubusercontent.com/ekoly/DS-Unit-2-Applied-Modeling/master/data/ufc/raw_total_fight_data.csv", sep=";")

dat_df.head()

R_fighter       B_fighter          Referee        date  \
0          Henry Cejudo   Marlon Moraes     Marc Goddard  2019-06-08   
1  Valentina Shevchenko     Jessica Eye  Robert Madrigal  2019-06-08   
2         Tony Ferguson  Donald Cerrone  Dan Miragliotta  2019-06-08   
3         Jimmie Rivera        Petr Yan  Kevin MacDonald  2019-06-08   
4           Tai Tuivasa   Blagoy Ivanov  Dan Miragliotta  2019-06-08   

                 location Winner  title_bout       weight_class  no_of_rounds  \
0  Chicago, Illinois, USA    Red        True       Bantamweight             5   
1  Chicago, Illinois, USA    Red        True  Women's Flyweight             5   
2  Chicago, Illinois, USA    Red       False        Lightweight             3   
3  Chicago, Illinois, USA   Blue       False       Bantamweight             3   
4  Chicago, Illinois, USA   Blue       False        Heavyweight             3   

   B_current_lose_streak  B_current_win_streak  B_draw  B_avg_BODY_att  \
0                    0.0                   4.0     0.0        9.200000   
1                    0.0                   3.0     0.0       14.600000   
2                    0.0                   3.0     0.0       15.354839   
3                    0.0                   4.0     0.0       17.000000   
4                    0.0                   1.0     0.0       17.000000   

   B_avg_BODY_landed  B_avg_CLINCH_att  B_avg_CLINCH_landed  \
0           6.000000          0.200000             0.000000   
1           9.100000         11.800000             7.300000   
2          11.322581          6.741935             4.387097   
3          14.000000         13.750000            11.000000   
4          14.500000          2.500000             2.000000   

   B_avg_DISTANCE_att  B_avg_DISTANCE_landed  B_avg_GROUND_att  \
0           62.600000              20.600000          2.600000   
1          124.700000              42.100000          2.400000   
2           84.741935              38.580645          5.516129   
3          109.500000              48.750000         13.000000   
4          201.000000              59.500000          0.000000   

   B_avg_GROUND_landed  B_avg_HEAD_att  B_avg_HEAD_landed  B_avg_KD  \
0             2.000000       48.600000          11.200000  0.800000   
1             1.900000      112.000000          32.000000  0.000000   
2             3.806452       67.645161          23.258065  0.645161   
3            10.500000      116.250000          53.750000  0.500000   
4             0.000000      184.500000          45.000000  0.000000   

   B_avg_LEG_att  B_avg_LEG_landed  B_avg_PASS  B_avg_REV  B_avg_SIG_STR_att  \
0            7.6          5.400000    0.400000   0.000000              65.40   
1           12.3         10.200000    0.800000   0.000000             138.90   
2           14.0         12.193548    0.935484   0.096774              97.00   
3            3.0          2.500000    0.500000   0.250000             136.25   
4            2.0          2.000000    0.000000   0.000000             203.50   

   B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  B_avg_SUB_ATT  B_avg_TD_att  \
0             22.600000           0.466000       0.400000       0.80000   
1             51.300000           0.399000       0.700000       1.00000   
2             46.774194           0.496129       0.354839       2.16129   
3             70.250000           0.550000       0.250000       2.50000   
4             61.500000           0.310000       0.000000       0.00000   

   B_avg_TD_landed  B_avg_TD_pct  B_avg_TOTAL_STR_att  B_avg_TOTAL_STR_landed  \
0         0.200000      0.100000            66.400000               23.600000   
1         0.500000      0.225000           158.700000               69.600000   
2         0.677419      0.295484           103.709677               52.548387   
3         1.250000      0.287500           154.750000               86.750000   
4         0.000000      0.000000           204.000000               62.000000   

   B_longest_win_streak  B_

In [12]:
# I'd like to have one row per fighter per fight. That is, 2 rows per fight, with a simple "is_winner" column.

import re


r_fighter_cols = [col for col in dat_df.columns if col.startswith("R_")]
b_fighter_cols = [col for col in dat_df.columns if col.startswith("B_")]

other_cols = [col for col in dat_df.columns if col not in r_fighter_cols and col not in b_fighter_cols]

def extractFighters(dat_df, fighter_cols, opp_cols, fighter_prefix, opp_prefix, fighter_color):

    fighters_df = dat_df[fighter_cols + other_cols].copy()
    fighters_df.columns = [col.replace(fighter_prefix, "") for col in fighters_df.columns]
    fighters_df["is_winner"] = fighters_df["Winner"] == fighter_color

    opp_df = dat_df[opp_cols].copy()
    opp_df.columns = [re.sub(r"^" + opp_prefix, "opponent_", col) for col in opp_df.columns]

    fighters_df = pd.concat([fighters_df, opp_df], axis=1)
    
    return fighters_df

r_fighters_df = extractFighters(dat_df, r_fighter_cols, b_fighter_cols, "R_", "B_", "Red")
b_fighters_df = extractFighters(dat_df, b_fighter_cols, r_fighter_cols, "B_", "R_", "Blue")

fighters_df = pd.concat([r_fighters_df, b_fighters_df], axis=0)
fighters_df.head()

Height_cms  Reach_cms          Referee    Stance  Weight_lbs Winner   age  \
0      162.56     162.56     Marc Goddard  Orthodox       135.0    Red  32.0   
1      165.10     167.64  Robert Madrigal  Southpaw       125.0    Red  31.0   
2      180.34     193.04  Dan Miragliotta  Orthodox       155.0    Red  35.0   
3      162.56     172.72  Kevin MacDonald  Orthodox       135.0   Blue  29.0   
4      187.96     190.50  Dan Miragliotta  Southpaw       264.0   Blue  26.0   

   avg_BODY_att  avg_BODY_landed  avg_CLINCH_att  avg_CLINCH_landed  \
0     21.900000        16.400000       17.000000          11.000000   
1     12.000000         7.714286        9.285714           6.857143   
2     13.866667         8.666667        2.866667           1.733333   
3     18.250000        10.250000        5.875000           4.125000   
4      7.750000         6.750000       11.000000           7.250000   

   avg_DISTANCE_att  avg_DISTANCE_landed  avg_GROUND_att  avg_GROUND_landed  \
0         75.000000            26.500000        9.400000           6.500000   
1         88.142857            36.142857       18.428571          16.428571   
2        116.133333            49.466667        5.333333           4.266667   
3        104.875000            41.000000        1.000000           0.625000   
4         50.750000            24.750000        0.500000           0.500000   

   avg_HEAD_att  avg_HEAD_landed  avg_KD  avg_LEG_att  avg_LEG_landed  \
0     74.200000            23.90   0.400     5.300000        3.700000   
1     84.571429            37.00   0.000    19.285714       14.714286   
2     96.733333            35.60   0.200    13.733333       11.200000   
3     80.500000            24.00   0.375    13.000000       11.500000   
4     50.750000            22.75   0.500     3.750000        3.000000   

   avg_PASS   avg_REV  avg_SIG_STR_att  avg_SIG_STR_landed  avg_SIG_STR_pct  \
0  1.200000  0.000000              NaN                 NaN              NaN   
1  1.714286  0.142857              NaN                 NaN              NaN   
2  0.333333  0.133333              NaN                 NaN              NaN   
3  0.125000  0.000000              NaN                 NaN              NaN   
4  0.250000  0.000000              NaN                 NaN              NaN   

   avg_SIG_STatt  avg_SIG_STlanded  avg_SIG_STpct  avg_SUATT  avg_SUB_ATT  \
0     101.400000         44.000000       0.466000        NaN     0.100000   
1     115.857143         59.428571       0.575714        NaN     0.428571   
2     124.333333         55.466667       0.430000        NaN     1.000000   
3     111.750000         45.750000       0.366250        NaN     0.000000   
4      62.250000         32.500000       0.545000        NaN     0.000000   

   avg_TD_att  avg_TD_landed  avg_TD_pct  avg_TOTAL_STR_att  \
0    5.300000       1.900000    0.458000                NaN   
1    5.142857       2.428571    0.601429                NaN   
2    0.933333       0.400000    0.277333                NaN   
3    2.250000       0.625000    0.103750                NaN   
4    0.500000       0.000000    0.000000                NaN   

   avg_TOTAL_STR_landed  avg_TOTAL_STatt  avg_TOTAL_STlanded  \
0                   NaN       129.900000           69.100000   
1                   NaN       161.571429          102.857143   
2                   NaN       133.000000           63.400000   
3                   NaN       117.375000           50.750000   
4                   NaN        63.500000           32.750000   

   avg_opp_BODY_att  avg_opp_BODY_landed  avg_opp_CLINCH_att  \
0         13.300000             8.800000            7.500000   
1         24.571429            14.142857           10.571429   
2         14.466667             8.133333            2.800000   
3         20.250000            13.375000            6.875000   
4          6.250000             4.750000            4.500000   

   avg_opp_CLINCH_landed  avg_opp_DISTANCE_att  avg_opp_DISTANCE_landed  \
0               5.1

In [13]:

for col in [col for col in fighters_df.select_dtypes(include="number").columns if col.startswith("opponent_")]:
    
    col2 = col
    col1 = col[9:]
    
    fighters_df[col1 + "_ratio"] = (fighters_df[col1] + 1) / (fighters_df[col2] + 1)
    
# prevent leakage
fighters_df = fighters_df.drop(columns="Winner")

# change prefixes to suffixes
fighters_df.columns = [col[9:] + "_opponent" if col.startswith("opponent_") else col for col in fighters_df.columns]


### Step 2: Establish Baseline

This is expected to be exactly 50% because for every winner there is a loser.

In [14]:
from sklearn.metrics import accuracy_score

majority_class = fighters_df["is_winner"].mode()[0]
baseline_pred = [majority_class] * len(fighters_df)
baseline_acc = accuracy_score(baseline_pred, fighters_df["is_winner"]) * 100

print(f"""
    Majority class: {majority_class}
    Baseline accuracy: {baseline_acc:.2f}%
""")


    Majority class: False
    Baseline accuracy: 50.81%



##### Explanation

It turns out 50.81% have False for "is_winner". This is probably because in the case of ties, neither fighter is considered the winner.

### Step 3: Linear Model

We will use a `LogisticRegression` and see if we can beat the baseline.

In [15]:
from sklearn.model_selection import train_test_split

from sklearn.pipeline import make_pipeline
import category_encoders as ce
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression


train, test = train_test_split(fighters_df)

target = "is_winner"
features = fighters_df.columns.drop([target] + ["age", "age_opponent", "Referee", "date"])

X_train = train[features]
y_train = train[target]

X_test = test[features]
y_test = test[target]

In [16]:
lr_pipeline = make_pipeline(
    ce.OneHotEncoder(use_cat_names=True),
    SimpleImputer(strategy="constant", fill_value="0"),
    LogisticRegression(max_iter=333),
)

lr_pipeline.fit(X_train, y_train)
y_pred = lr_pipeline.predict(X_test)

lr_score = accuracy_score(y_pred, y_test)*100
improvement = lr_score - baseline_acc

print(f"""
    Accuracy score for LogisticRegression: {lr_score:.2f}%
    Improvement over baseline: {improvement:.2f}%
""")

/home/cowley/venv/lambda/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:939: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html.
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)



    Accuracy score for LogisticRegression: 65.36%
    Improvement over baseline: 14.55%



We achieved a notable improvement over baseline with a logistic regression.

### Step 4: Try RandomForestClassifier

We will use a `RandomForestClassifier` and see if we can beat the logistic regression.

In [17]:
from sklearn.ensemble import RandomForestClassifier

rfc_pipeline = make_pipeline(
    ce.OneHotEncoder(use_cat_names=True),
    SimpleImputer(strategy="constant", fill_value=0),
    RandomForestClassifier(n_jobs=-1, n_estimators=220),
)

rfc_pipeline.fit(X_train, y_train)
y_pred = rfc_pipeline.predict(X_test)

rfc_score = accuracy_score(y_pred, y_test)*100
improvement_base = rfc_score - baseline_acc
improvement_lr = rfc_score - lr_score

print(f"""
    Accuracy score for Random Forest: {rfc_score:.2f}%
    Improvement over baseline: {improvement_base:.2f}%
    Improvement over logistic: {improvement_lr:.2f}%
""")


    Accuracy score for Random Forest: 66.76%
    Improvement over baseline: 15.95%
    Improvement over logistic: 1.40%



### Step 5: Try XGBClassifier

We will use a `XGBClassifier` to see if we can beat the random forest.

The parameters were optimized with a `RandomizedSearchCV` in a separate notebook.

In [18]:
from xgboost import XGBClassifier

xgb_pipeline = make_pipeline(
    ce.OneHotEncoder(use_cat_names=True),
    SimpleImputer(strategy="constant", fill_value=0),
    XGBClassifier(
        n_jobs=-1,
        n_estimators=220,
        max_depth=76,
        learning_rate=0.22,
    )
)


xgb_pipeline.fit(X_train, y_train)
y_pred = xgb_pipeline.predict(X_test)

xgb_score = accuracy_score(y_pred, y_test)*100
improvement_base = xgb_score - baseline_acc
improvement_lr = xgb_score - lr_score
improvement_rfc = xgb_score - rfc_score

print(f"""
    Accuracy score for XGB: {xgb_score:.2f}%
    Improvement over baseline: {improvement_base:.2f}%
    Improvement over logistic: {improvement_lr:.2f}%
    Improvement over random forest: {improvement_rfc:.2f}%
""")


    Accuracy score for XGB: 66.95%
    Improvement over baseline: 16.15%
    Improvement over logistic: 1.59%
    Improvement over random forest: 0.19%



### Step 6: Permutation Importance

Analyze which features may detract from the accuracy.

In [19]:
from eli5.sklearn import PermutationImportance

transformer = make_pipeline(
    ce.OrdinalEncoder(),
    SimpleImputer(strategy="constant", fill_value=0),
)

X_train_transformed = transformer.fit_transform(X_train)
X_test_transformed = transformer.transform(X_test)

/home/cowley/venv/lambda/lib/python3.6/site-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.metrics.scorer module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.metrics. Anything that cannot be imported from sklearn.metrics is now part of the private API.
  warnings.warn(message, FutureWarning)
/home/cowley/venv/lambda/lib/python3.6/site-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.feature_selection.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.feature_selection. Anything that cannot be imported from sklearn.feature_selection is now part of the private API.
  warnings.warn(message, FutureWarning)


In [20]:
from matplotlib import pyplot as plt

model = XGBClassifier(
    n_jobs=-1,
    n_estimators=220,
    max_depth=76,
    learning_rate=0.22,
)
model.fit(X_train_transformed, y_train)

permuter = PermutationImportance(
    model,
    scoring="accuracy",
    n_iter=5,
)

permuter.fit(X_test_transformed, y_test)

PermutationImportance(cv='prefit',
                      estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                              colsample_bylevel=1,
                                              colsample_bynode=1,
                                              colsample_bytree=1, gamma=0,
                                              learning_rate=0.22,
                                              max_delta_step=0, max_depth=76,
                                              min_child_weight=1, missing=None,
                                              n_estimators=220, n_jobs=-1,
                                              nthread=None,
                                              objective='binary:logistic',
                                              random_state=0, reg_alpha=0,
                                              reg_lambda=1, scale_pos_weight=1,
                                              seed=None, silent=None,
                  

In [ ]:
feature_importances_plus_std = pd.Series(permuter.feature_importances_ + permuter.feature_importances_std_, features)

xgb_best_features = fighters_df.columns.drop([target] + ["age", "age_opponent", "Referee", "date"] + list(feature_importances_plus_std[feature_importances_plus_std <= 0].index))

X_train_xgb = train[xgb_best_features]
X_test_xgb = test[xgb_best_features]

xgb_pipeline = make_pipeline(
    ce.OneHotEncoder(use_cat_names=True),
    SimpleImputer(strategy="constant", fill_value=0),
    XGBClassifier(
        n_jobs=-1,
        n_estimators=220,
        max_depth=76,
        learning_rate=0.22,
    )
)


xgb_pipeline.fit(X_train_xgb, y_train)
y_pred = xgb_pipeline.predict(X_test_xgb)

xgb_score2 = accuracy_score(y_pred, y_test)*100
improvement_base = xgb_score2 - baseline_acc
improvement_xgb = xgb_score2 - xgb_score

print(f"""
    Accuracy score for XGB with optimized features: {xgb_score2:.2f}%
    Improvement over baseline: {improvement_base:.2f}%
    Improvement over base XGB: {improvement_xgb:.2f}%
""")

Marginal gains.

Let's try again with Random Forest.

In [ ]:
model = RandomForestClassifier(
    n_jobs=-1,
    n_estimators=220,
)
model.fit(X_train_transformed, y_train)

permuter = PermutationImportance(
    model,
    scoring="accuracy",
    n_iter=5,
)

permuter.fit(X_test_transformed, y_test)

In [ ]:
feature_importances_plus_std = pd.Series(permuter.feature_importances_ + permuter.feature_importances_std_, features)

rfc_best_features = fighters_df.columns.drop([target] + ["age", "age_opponent", "Referee", "date"] + list(feature_importances_plus_std[feature_importances_plus_std <= 0].index))

X_train_rfc = train[rfc_best_features]
X_test_rfc = test[rfc_best_features]

rfc_pipeline = make_pipeline(
    ce.OneHotEncoder(use_cat_names=True),
    SimpleImputer(strategy="constant", fill_value=0),
    RandomForestClassifier(
        n_jobs=-1,
        n_estimators=220,
    )
)

rfc_pipeline.fit(X_train_rfc, y_train)
y_pred = rfc_pipeline.predict(X_test_rfc)

rfc_score2 = accuracy_score(y_pred, y_test)*100
improvement_base = rfc_score2 - baseline_acc
improvement_rfc = rfc_score2 - rfc_score

print(f"""
    Accuracy score for Random Forest with optimized features: {rfc_score2:.2f}%
    Improvement over baseline: {improvement_base:.2f}%
    Improvement over base RFC: {improvement_rfc:.2f}%
""")

To wrap up the permutations, I'd like to make two more visualizations showing the feature importances.

In [ ]:
feature_importances = pd.Series(permuter.feature_importances_, features).sort_values()
feature_importances_std = pd.Series(permuter.feature_importances_std_, features).sort_values()

good_mask = feature_importances > 0
bad_mask = feature_importances <= 0

fig, ax = plt.subplots(figsize=(8, 30))

feature_importances[good_mask].plot.barh(color="green", ecolor="#105010", xerr=feature_importances_std[good_mask], ax=ax)
ax.set_title("Positive Features")

fig.show()

In [ ]:
fig, ax = plt.subplots(figsize=(8, 30))

feature_importances[bad_mask].plot.barh(color="red", ecolor="#501010", xerr=feature_importances_std[bad_mask], ax=ax)
ax.set_title("Negative Features")

fig.show()

### Run More Tests

While the previous tests were purely numbers-based, the following human-oriented tests will predict the outcome of real high-profile matches. 

In [ ]:


prefix_re = re.compile(r".*(_opponent)|(_ratio)$")

fighters_individual_df = fighters_df[[col for col in fighters_df.columns.drop(["is_winner"]) if not prefix_re.match(col)]]
latest_fights = fighters_individual_df.sort_values(by="date").groupby("fighter").tail(1)

def getByFighter(fighter_name):
    return latest_fights[latest_fights["fighter"] == fighter_name].copy()

def makeBoutDf(fighter1, fighter2):
    
    fighter1_df = getByFighter(fighter1)
    fighter2_df = getByFighter(fighter2)
    
    fighter1_df["temp_id_"] = fighter2_df["temp_id_"] = np.random.randint(2**31)
    fighter1_df, fighter2_df = (
        pd.merge(fighter1_df, fighter2_df, on="temp_id_", suffixes=("", "_opponent")),
        pd.merge(fighter2_df, fighter1_df, on="temp_id_", suffixes=("", "_opponent")),
    )
    
    fight_df = pd.concat([fighter1_df, fighter2_df])
    
    for col in [col for col in fight_df.select_dtypes(include="number").columns if col.endswith("_opponent")]:

        col2 = col
        col1 = col[:-9]

        fight_df[col1 + "_ratio"] = (fight_df[col1] + 1) / (fight_df[col2] + 1)
    
    fight_df["stance_config"] = fight_df["Stance"] + "-" + fight_df["Stance_opponent"]
    
    return fight_df.drop(columns=["temp_id_"])

def scoreBout(bout):
    return pd.DataFrame(data=pipeline.predict_proba(bout[features]), columns=[str(x) for x in pipeline.classes_])

def doPrediction(red_fighter, blue_fighter):
    bout = makeBoutDf(red_fighter, blue_fighter)
    probas = scoreBout(bout)
    
    red_fighter_prob = (probas.iloc[0]["True"] + probas.iloc[1]["False"])/2
    blue_fighter_prob = (probas.iloc[0]["False"] + probas.iloc[1]["True"])/2
    
    if red_fighter_prob > blue_fighter_prob:
        return red_fighter_prob, red_fighter
    
    else:
        return blue_fighter_prob, blue_fighter

trials = [
    ("Kamaru Usman", "Colby Covington"),
    ("Jorge Masvidal", "Nate Diaz"),
    ("Conor McGregor", "Khabib Nurmagomedov"),
    ("Robert Whittaker", "Israel Adesanya"),
    ("Max Holloway", "Frankie Edgar"),
    ("Michelle Waterson", "Joanna Jedrzejczyk"),
    ("Amanda Nunes", "Germaine de Randamie"),
    ("Jon Jones", "Anthony Smith"),
    ("Khabib Nurmagomedov", "Dustin Poirier"),
    ("Ben Askren", "Robbie Lawler"),
    ("Alexander Volkanovski", "Max Holloway"),
    ("Ben Askren", "Jorge Masvidal"),
]

def runTests():
    for b_fighter, r_fighter in trials:
        print(b_fighter, "vs", r_fighter)
        print("\t", doPrediction(b_fighter, r_fighter))

In [ ]:
pipeline = rfc_pipeline
features = rfc_best_features
runTests()

In [ ]:
pipeline = xgb_pipeline
features = xgb_best_features
runTests()

### Export the best pipeline

In [ ]:
import pickle

with open("../pickles/pipeline.pickle", "wb") as f:
    pickle.dump(xgb_pipeline, f)
    
with open("../pickles/features.pickle", "wb") as f:
    pickle.dump(xgb_best_features, f)


### More Visualizations

In [ ]:
from plotly.offline import init_notebook_mode, iplot, plot
from plotly import express as px

init_notebook_mode(connected=True)

fig = px.scatter(fighters_df, x="Reach_cms", y="age_ratio", color="is_winner")
plot(fig)

In [10]:
fighters_df[fighters_df["fighter"] == "Khabib Nurmagomedov"]

Height_cms  Reach_cms          Referee    Stance  Weight_lbs   age  \
347        177.8      177.8        Herb Dean  Orthodox       155.0  30.0   
585        177.8      177.8  Dan Miragliotta  Orthodox       155.0  29.0   
698        177.8      177.8    John McCarthy  Orthodox       155.0  29.0   
1251       177.8      177.8    John McCarthy  Orthodox       155.0  28.0   
1519       177.8      177.8    James Warring  Orthodox       155.0  27.0   
2858       177.8      177.8   Mario Yamasaki  Orthodox       155.0  24.0   
2478       177.8      177.8    John McCarthy  Orthodox       155.0  25.0   
2735       177.8      177.8     Yves Lavigne  Orthodox       155.0  25.0   
2995       177.8      177.8  Dan Miragliotta  Orthodox       155.0  24.0   
3162       177.8      177.8   Mario Yamasaki  Orthodox       155.0  23.0   
3329       177.8      177.8   Mario Yamasaki  Orthodox       155.0  23.0   

      avg_BODY_att  avg_BODY_landed  avg_CLINCH_att  avg_CLINCH_landed  \
347       6.400000         4.400000        3.600000           2.500000   
585       5.777778         4.111111        4.000000           2.777778   
698       5.125000         3.625000        4.125000           2.875000   
1251      5.714286         4.000000        4.571429           3.142857   
1519      6.666667         4.666667        5.166667           3.500000   
2858      6.000000         3.333333        5.333333           4.333333   
2478      6.400000         4.400000        5.400000           3.400000   
2735      5.500000         3.250000        5.000000           3.750000   
2995      8.000000         5.000000        8.000000           6.500000   
3162      4.000000         3.000000        4.000000           3.000000   
3329           NaN              NaN             NaN                NaN   

      avg_DISTANCE_att  avg_DISTANCE_landed  avg_GROUND_att  \
347          72.400000            23.900000       44.400000   
585          52.111111            15.444444       43.888889   
698          46.625000            13.375000       39.625000   
1251         50.285714            14.571429       28.142857   
1519         57.000000            16.500000       17.166667   
2858         49.333333            11.666667       15.000000   
2478         54.000000            17.000000       16.400000   
2735         39.750000            10.250000       17.500000   
2995         68.000000            15.000000       13.000000   
3162         47.000000            15.000000       26.000000   
3329               NaN                  NaN             NaN   

      avg_GROUND_landed  avg_HEAD_att  avg_HEAD_landed    avg_KD  avg_LEG_att  \
347           32.300000    110.600000        51.200000  0.200000     3.400000   
585           32.111111     91.000000        43.333333  0.222222     3.222222   
698           29.250000     82.125000        39.125000  0.250000     3.125000   
1251          20.857143     73.714286        31.428571  0.285714     3.571429   
1519          12.500000     68.833333        24.500000  0.333333     3.833333   
2858          11.333333     61.000000        22.333333  0.666667     2.666667   
2478          11.400000     66.000000        24.600000  0.400000     3.400000   
2735          12.250000     53.750000        20.750000  0.500000     3.000000   
2995           8.500000     77.000000        22.500000  0.500000     4.000000   
3162          17.000000     72.000000        32.000000  1.000000     1.000000   
3329                NaN           NaN              NaN       NaN          NaN   

      avg_LEG_landed  avg_PASS  avg_REV  avg_SIG_STR_att  avg_SIG_STR_landed  \
347         3.100000  4.500000      0.0              NaN                 NaN   
585         2.888889  4.222222      0.0              NaN                 NaN   
698         2.750000  4.250000      0.0              NaN                 NaN   
1251        3.142857  4.000000      0.0              NaN                 NaN   
1519        3.333333  3.500000      0.0              NaN                 N